src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -117,45 +130,69 @@ public void Dispose()
             await Task.Yield();
 
             _logger.LogTrace("Beginning receive loop");
-            while (!cancellationToken.IsCancellationRequested)
+            try
             {
-                // This is a little odd... we want to remove the InvocationRequest once and only once so we pull it out in the callback,
-                // and stash it here because we know the callback will have finished before the end of the await.
-                var message = await _adapter.ReadMessageAsync(_stream, _binder, cancellationToken);
-
-                var invocationDescriptor = message as InvocationDescriptor;
-                if (invocationDescriptor != null)
-                {
-                    DispatchInvocation(invocationDescriptor, cancellationToken);
-                }
-                else
+                while (!cancellationToken.IsCancellationRequested)
                 {
-                    var invocationResultDescriptor = message as InvocationResultDescriptor;
-                    if (invocationResultDescriptor != null)
+                    // This is a little odd... we want to remove the InvocationRequest once and only once so we pull it out in the callback,
+                    // and stash it here because we know the callback will have finished before the end of the await.
+                    var message = await _adapter.ReadMessageAsync(_stream, _binder, cancellationToken);
+
+                    var invocationDescriptor = message as InvocationDescriptor;
+                    if (invocationDescriptor != null)
+                    {
+                        DispatchInvocation(invocationDescriptor, cancellationToken);
+                    }
+                    else
                     {
-                        DispatchInvocationResult(invocationResultDescriptor, cancellationToken);
+                        var invocationResultDescriptor = message as InvocationResultDescriptor;
+                        if (invocationResultDescriptor != null)
+                        {
+                            InvocationRequest irq;
+                            lock (_pendingCallsLock)
+                            {
+                                _connectionActive.Token.ThrowIfCancellationRequested();
+                                irq = _pendingCalls[invocationResultDescriptor.Id];
+                                _pendingCalls.Remove(invocationResultDescriptor.Id);
+                            }
+                            DispatchInvocationResult(invocationResultDescriptor, irq, cancellationToken);
+                        }
                     }
                 }
             }
-            _logger.LogTrace("Ending receive loop");
+            finally
```

Comment:
We should review and improve exception handling for sure. It's only a problem if an exception occurs that did **not** come from the pipeline, since we have the ContinueWith. I don't disagree that we should be better at protecting against exceptions here though. I'll file an issue for improved exception handling and testing.